### PR TITLE
PHP 8 Support

### PR DESCRIPTION
**Pull request by driesvints**
Working on providing full PHP 8 support for php-webdriver. We'll need to for Laravel Dusk.

Currently using this PR to run the builds.

### CODE REVIEW

File: phpunit.xml.dist
Comment (williamdes) [inline phpunit.xml.dist:21]: This change is not necessary and is not good because we still support phpunit 5 ;)

Comment (driesvints) [reply phpunit.xml.dist:21]: Yeah, I'm trying to get the PHP 8 build to pass first. I'll revert everything that's unnecessary of course :thu

**Comment by driesvints** [comment]
Currently waiting on https://github.com/sminnee/phpunit-mock-objects/pull/6 to be merged and tagged before continuing here.

File: phpunit.xml.dist
Comment (sminnee) [reply phpunit.xml.dist:21]: We had this issue on Silverstripe and I addressed this by creating sminnee/phpunit, a replacement that ships 5.7.30 with PHP8 support.

Ditto for sminnee/phpunit-mock-objects

You're welcome to use it if you want.

Comment (driesvints) [reply phpunit.xml.dist:21]: Thanks @sminnee, I'll take a look at everything on Monday 👍

**Comment by sminnee** [comment]
You may wish to run your php 8 build with --ignore-platform-reqs. Without dropping PHP 5.6 and 7.0 support, any upgrading to at least phpunit 7 || 8 || 9, you'll likely run into issues.

If you're looking to patch my phpunit forks, please test the new packages against phpunit/phpunit first, and confirm that the tests pass - this is what I got blocked on trying to upgrade some of the dependencies.

File: phpunit.xml.dist
Comment (driesvints) [reply phpunit.xml.dist]: Reverted this. And implemented sminnee/phpunit, thanks!

**Comment by driesvints** [comment]
@sminnee I've added "sminnee/phpunit" which indeed works thanks.

I'm a bit stuck atm... For some bizar reason when I only run the build for PHP 8 it succeeds (see [this build](https://github.com/php-webdriver/php-webdriver/runs/1342588938) and my second to last commit) but when I re-enable all the other builds it fails and claims it cannot install PHP CS Fixer even though I've added `--ignore-platform-req php` (see [this build](https://travis-ci.com/github/php-webdriver/php-webdriver/jobs/424113941) and my last commit). Apparently for some reason it seems that parallel builds in Travis influence each other somehow? In any case the PHP 8 nightly build should properly install because of the `--ignore-platform-req php` flag.

**Comment by driesvints** [comment]
Also the PHP Code Style Github Action fails atm but I suspect that has nothing to do with the current PR (since it's run on PHP 7.4). I believe when you trigger it on main it'll also fail. Probably something got changed on phpstan's side.

**Comment by driesvints** [comment]
I think I'm a little closer at figuring it out. As you can see below there's two nightly builds, one that fails and one that passes. I'm sure that the one that passes has the `--ignore-platform-req php` and the one that fails doesn't.

There's two questions:

- Why are there two builds?
- How do we get the `--ignore-platform-req php` to apply on the one that fails?

Screenshot 2020-11-02 at 17 37 30

File: .travis.yml
Comment (williamdes) [inline .travis.yml:11]: Remove this one and you will not have 2 builds ;)

Comment (driesvints) [reply .travis.yml]: That seem to have done it. Thank you @williamdes! I guess I was confused here because if you leave it in place but disable all the other matrix builds it only runs one. That's pretty confusing...

Comment (williamdes) [reply .travis.yml]: In fact if you add this one here it will run the matrix job for each value
And you added a special block for a job and that made it to be duplicated.
Here is an example of such override https://github.com/phpmyadmin/sql-parser/blob/37c4baeddee8f749b0ce3a26e88988c5d39dfa60/.travis.yml#L42

**Comment by driesvints** [comment]
Okay this one is ready for review now. Please note that I had to disable code coverage for PHP 8 for now because of an issue with PHPUnit and xdebug. See https://travis-ci.com/github/php-webdriver/php-webdriver/builds/197177332

This was fixed in a newer version of PHPUnit later on but I guess `sminnee/phpunit` doesn't has that one yet (cc @sminnee). I'm however a bit too tired to figure out that specific one and don't believe it should be a blocker for this PR.

In general I want to suggest for this project to drop support for older PHP versions so you can remove the PHPUnit dependency overrides. As most of these PHP versions are EoL I don't think it makes sense in maintaining them anymore. However, I understand that this is a personal choice of the maintainer of course.

Anyway, would appreciate a review, merge and tag so we can finally patch Laravel Dusk. Let me know if anything else needs to be done. Thanks all! 🙂

File: .travis.yml
Comment (williamdes) [inline .travis.yml:126]: is this temporary ?

File: composer.json
Comment (williamdes) [inline composer.json:31]: we should stick to `phpunit/phpunit` and add new supported versions if possible.

File: .travis.yml
Comment (williamdes) [reply .travis.yml:126]: --no-coverage could be enough to work

File: composer.json
Comment (driesvints) [reply composer.json:31]: That's not possible because `sminnee/phpunit-mock-objects` is blocking that atm. See https://github.com/sminnee/phpunit-mock-objects/pull/6 and his reply (and his other replies on this PR).

**Comment by driesvints** [comment]
For some bizar reason the PHP 5.6 til 7.2 builds are now failing after my last commit... https://github.com/php-webdriver/php-webdriver/pull/828/checks?check_run_id=1343276205

File: composer.json
Comment (williamdes) [reply composer.json:31]: Okay, I can finish your work then ;)
Please give me access to your repository

Comment (driesvints) [reply composer.json:31]: Not sure which one you meant so I gave you access to this one and the phpunit-mock-objects fork :)

File: .travis.yml
Comment (driesvints) [reply .travis.yml]: This worked 👍

File: .travis.yml
Comment (driesvints) [reply .travis.yml]: Right. But I think the current state of the PR is good enough?

File: composer.json
Comment (williamdes) [reply composer.json:31]: Thank you, perfect :rocket: 
This was what I expected :)

**Comment by williamdes** [comment]
Part of the build passes, I will copy my hack friends into the CI to make tests pass on modern versions of phpunit ;)
Source: https://salsa.debian.org/phpmyadmin-team/phpmyadmin/-/blob/debian/4%254.9.7+dfsg1-1_bpo10+1/debian/tests/phpunit#L17

File: .php_cs.dist
Comment (williamdes) [inline .php_cs.dist:55]: Because assertInternalType is deprecated

**Comment by williamdes** [comment]
Hi @OndraM !
This PR started with a PHP 8 support and I finished it with a phpunit 5.7->9 support.
Because only the CI runs phpunit tests on PHP 5 I did set a script block to reverse some method signatures like I do on Debian (https://github.com/php-webdriver/php-webdriver/pull/828#issuecomment-720622743).

I upgraded things and did fix everything that was red.
Some fixes/hacks are needed and can not be removed. But we can merge this and remove those things when we drop php 5 support.

**Comment by driesvints** [comment]
@williamdes fantastic work 🙌

File: .php_cs.dist
Comment (GrahamCampbell) [reply .php_cs.dist:55]: Just delete this line.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by williamdes** [comment]
Hi @GrahamCampbell 
> Just delete this line.

I can not do that because phpcs will report me to use some phpunit asserts that are not right.
See: 26c6de1cdb8575ac4d06f8c38530bb11c97d644e
Deprecated: https://github.com/sebastianbergmann/phpunit/issues/3369

**Review by mauriciofauth — APPROVED** [review]

File: tests/unit/Interactions/Internal/WebDriverCoordinatesTest.php
Comment (OndraM) [inline tests/unit/Interactions/Internal/WebDriverCoordinatesTest.php:17]: This test does seem to focus on internal implementation... I know its unrelated to PHP 8, but instead of using reflectioins, we can rather replace it with something more useful, so I suggest replacing  whole method it with following:

```php
public function testConstruct()
    {
        $in_view_port = function () {
            return new WebDriverPoint(0, 0);
        };
        $on_page = function () {
            return new WebDriverPoint(10, 10);
        };

        $webDriverCoordinates = new WebDriverCoordinates(null, $in_view_port, $on_page, 'auxiliary');

        $this->assertEquals(new WebDriverPoint(0, 0), $webDriverCoordinates->inViewPort());
        $this->assertEquals(new WebDriverPoint(10, 10), $webDriverCoordinates->onPage());
        $this->assertSame('auxiliary', $webDriverCoordinates->getAuxiliary());
    }
```

Then we can also drop following "testGetAuxiliary" method.

File: .travis.yml
Comment (OndraM) [inline .travis.yml:96]: What about moving this to separate shell script to make the config shorter?

Also we may eg. soon migrate to GH Actions only, so there won't be a need to copy-paste it.

File: tests/unit/Interactions/Internal/WebDriverCoordinatesTest.php
Comment (driesvints) [reply tests/unit/Interactions/Internal/WebDriverCoordinatesTest.php:19]: Done 👍

File: tests/functional/WebDriverTestCase.php
Comment (OndraM) [inline tests/functional/WebDriverTestCase.php:223]: Just in idea, isn't possible to name it simply "assertStringContainsString", and call the parent if it exists, or assertContains if not?

Once we drop old PHPUnit versions, we could then only remove this polyfill.

Comment (driesvints) [reply tests/functional/WebDriverTestCase.php:223]: @williamdes can you maybe check in on this one?

File: .travis.yml
Comment (driesvints) [reply .travis.yml:96]: @williamdes can you maybe check in on this one?

File: tests/functional/WebDriverTestCase.php
Comment (williamdes) [reply tests/functional/WebDriverTestCase.php:223]: No I can not do that, neither.
Because if I rename to `compatAssertStringContainsString` I will have to be compatible on the PHP signature, that would be one particular phpunit version. And so I would screw all the other ones.
Some not supporting phptype hints, some having some typehints defined, some no typehints defined.
:man_shrugging:

Comment (OndraM) [reply tests/functional/WebDriverTestCase.php:223]: Though the signature removal is already being done in via `sed` for similar cases like `setUpBeforeClass` etc., so theoretically it could be possible for `assertStringContainsString` signature as well, no? 

However it was just an idea, if you think its not worth it, we can keep it like this :).

Comment (williamdes) [reply tests/functional/WebDriverTestCase.php:223]: Well it is possible but totally overkill, let's keep it as it is ;)
I say it is overkill because we would need to make different seds for different phpunit versions and this is more complicated than just switching `: void` return typehints.
So is this okay for you until we drop some phpunit versions ?

File: .travis.yml
Comment (williamdes) [reply .travis.yml]: Done in 720c2750d4eb3064e6363f3f7a7c16fba75cca59

Comment (driesvints) [reply .travis.yml]: Just a really minor thing but maybe name the directory `bin/` so it's more in line with other composer libraries?

Comment (williamdes) [reply .travis.yml]: I would say no, on [Doctum](https://github.com/code-lts/doctum/tree/main/bin) and some other libs at [phpMyAdmin](https://github.com/phpmyadmin/sql-parser/tree/master/bin) scripts found in bin are for users of the lib :)

Comment (driesvints) [reply .travis.yml]: Gotcha 👍

**Comment by OndraM** [comment]
LGTM, thanks everyone!

The final question is how to rebase & fixup those 25 commits to keep the changes contained. Any preferences @driesvints  @williamdes?
I suggest keeping just two commits, one which adds Travis builds, second one with all the changes from @williamdes :).

**Comment by driesvints** [comment]
@OndraM all good to me. I'll leave it up to @williamdes to decide on that since he did the bulk of the work here 👍

**Comment by williamdes** [comment]
> @OndraM all good to me. I'll leave it up to @williamdes to decide on that since he did the bulk of the work here

Done, I did a reset soft and created 2 commits from the changes.
And added `Dries Vints <dries@vints.io>` as a co author on 06ff635be7be7fc3ca20f8f4bf2406e7324236a3

**Comment by driesvints** [comment]
@williamdes can you try to retrigger the static analysis github action? It might just be a hiccup.

**Comment by williamdes** [comment]
Hi @ondrejmirtes
This phpstan error seems to be recent, is it an internal bug ?
https://github.com/php-webdriver/php-webdriver/runs/1406254306?check_suite_focus=true#step:6:28

**Comment by ondrejmirtes** [comment]
@williamdes https://github.com/phpstan/phpstan/issues/3956

**Comment by OndraM** [comment]
So maybe we can temporarily add constrain to composer.json to avoid latest PHPStan version(s)?

**Comment by ondrejmirtes** [comment]
It'd be easier to:

1) Avoid using non-UTF-8 PHP code in your project. It's 2020 :)
2) The root issue is reported here (https://github.com/clue/reactphp-ndjson/issues/21). I don't know what to do about it right now. I don't know anything about the encoding of the analysed codebse so I don't know what to convert it from to UTF-8.

**Comment by OndraM** [comment]
@ondrejmirtes AFAIK this whole codebase is UTF-8. Is there a way how to find out which file causes the trouble?

I also tried this and find no non-UTF-8 files... https://gist.github.com/rmetzler/2947828

**Comment by williamdes** [comment]
I tried on my local workstation:
```
composer analyse
# fails
vendor/bin/phpstan.phar analyze ./lib ./tests --level 2 -c phpstan.neon --ansi
# fails
vendor/bin/phpstan.phar analyze ./lib ./tests --level 2 -c phpstan.neon --ansi
# fails
vendor/bin/phpstan.phar analyze ./lib ./tests --level 2 -c phpstan.neon --ansi --debug -vvv
# SUCCESS
vendor/bin/phpstan.phar analyze ./lib ./tests --level 2 -c phpstan.neon --ansi
# SUCCESS, WTF
rm -rf /tmp/phpstan
vendor/bin/phpstan.phar analyze ./lib ./tests --level 2 -c phpstan.neon --ansi
# fails
```

@ondrejmirtes `--debug -vvv` was added in the latest release, and before this release the analysis was successing.
Is it by coincidence this new error ?

**Comment by ondrejmirtes** [comment]
The reason why it fails are these two commits:

* https://github.com/phpstan/phpstan-src/commit/95101a01350698ecaa447fbf791516b8f0a5ccee
* https://github.com/phpstan/phpstan-src/commit/14928bf41fbe80ba68ab484e138138442d1cef4a

So it's not a new bug, it just surfaces an error condition that was failing silently before, possibly causing error reports to be missed.

**Comment by driesvints** [comment]
@OndraM @williamdes is there any way we can work around this for now so we can merge this PR? Afaik this has nothing to do with this PR and it'll probably also fail on the main branch?

**Comment by OndraM** [comment]
From what I found, the cause is in [WebDriveKeys](https://github.com/php-webdriver/php-webdriver/blob/main/lib/WebDriverKeys.php#L12-L81). However we need to declare WebDriver key codes there... Doing this is nothing invalid from PHP point of view. @ondrejmirtes So this is caused by https://github.com/clue/reactphp-ndjson/issues/21 , which is used by PHPStan, and which causes non-unicode characters to break JSON parsing used internally in PHPStan?

**Comment by OndraM** [comment]
> @OndraM @williamdes is there any way we can work around this for now so we can merge this PR? Afaik this has nothing to do with this PR and it'll probably also fail on the main branch?

Yes, will probably do this. I just wanted to know what is the root cause for this and whether it is related to this PR.

**Comment by OndraM** [comment]
Workaround for the PHPStan issue here: https://github.com/php-webdriver/php-webdriver/pull/830

File: phpstan.neon
Comment (OndraM) [inline phpstan.neon:11]: And now those needs to be put back :upside_down_face: .

Comment (williamdes) [reply phpstan.neon]: Thank you !
Happy you did point that out because I was searching why the error was occurring ^^

**Comment by driesvints** [comment]
@ondrejmirtes @OndraM @williamdes thanks all! 🙂

**Comment by OndraM** [comment]
PHP 8 support is now merged to `main` branch. Nice job everyone! 

I will get this released as `1.8.4` quite soon.

**Comment by viezel** [comment]
@ondrejmirtes @OndraM @williamdes awesome work. Looking forward to that `1.9.0` tag! 
PHP8 here we go! 🚀

**Comment by driesvints** [comment]
Hey @OndraM. Is there anything blocking the 1.9.0? We'd really like to patch Dusk soon since it's currently blocking a lot of other projects from shipping PHP 8 support. Let me know if there's anything I can do to help out.

**Comment by OndraM** [comment]
@driesvints Sorry for the delay, I'd like to finish one other PR (#556) to get it merged into the 1.9.0 release. Should be finished once CI builds are done, and the release will be just afterwards :)

**Comment by driesvints** [comment]
@OndraM no sorry needed. Awesome, looking forward to it :)

**Comment by OndraM** [comment]
:loudspeaker: 1.9.0 with PHP 8 support was just released.

**Comment by driesvints** [comment]
@OndraM amazing!

File: scripts/apply-phpunit-patches.sh
Comment (driesvints) [inline scripts/apply-phpunit-patches.sh:20]: @OndraM @williamdes would it be okay with you two if I used this script for https://github.com/algolia/algoliasearch-client-php/pull/644 as well?

Comment (williamdes) [reply scripts/apply-phpunit-patches.sh:20]: Yes you can copy it :)

Comment (OndraM) [reply scripts/apply-phpunit-patches.sh:20]: 👍

**Pull request merged**
